substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,763 @@
+			if let Some(existing_delegation) = Delegation::<T>::get(delegator) {
+				ensure!(&existing_delegation.agent == agent, Error::<T>::InvalidDelegation);
+				existing_delegation
+					.amount
+					.checked_add(&amount)
+					.ok_or(ArithmeticError::Overflow)?
+			} else {
+				amount
+			};
+
+		Delegation::<T>::new(agent, new_delegation_amount).update_or_kill(delegator);
+		ledger.total_delegated =
+			ledger.total_delegated.checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+		ledger.update(agent);
+
+		Self::deposit_event(Event::<T>::Delegated {
+			agent: agent.clone(),
+			delegator: delegator.clone(),
+			amount,
+		});
+
+		Ok(())
+	}
+
+	fn do_release(
+		who: &T::AccountId,
+		delegator: &T::AccountId,
+		amount: BalanceOf<T>,
+		num_slashing_spans: u32,
+	) -> DispatchResult {
+		let mut agent = Agent::<T>::get(who)?;
+		let mut delegation = Delegation::<T>::get(delegator).ok_or(Error::<T>::NotDelegator)?;
+
+		// make sure delegation to be released is sound.
+		ensure!(&delegation.agent == who, Error::<T>::NotAgent);
+		ensure!(delegation.amount >= amount, Error::<T>::NotEnoughFunds);
+
+		// if we do not already have enough funds to be claimed, try withdraw some more.
+		if agent.ledger.unclaimed_withdrawals < amount {
+			// withdraw account.
+			let _ = T::CoreStaking::withdraw_unbonded(who.clone(), num_slashing_spans)
+				.map_err(|_| Error::<T>::WithdrawFailed)?;
+			// reload agent from storage since withdrawal might have changed the state.
+			agent = agent.refresh()?;
+		}
+
+		// if we still do not have enough funds to release, abort.
+		ensure!(agent.ledger.unclaimed_withdrawals >= amount, Error::<T>::NotEnoughFunds);
+
+		// claim withdraw from agent. Kill agent if no delegation left.
+		// TODO(ank4n): Ideally if there is a register, there should be an unregister that should
+		// clean up the agent. Need to improve this.
+		if agent.remove_unclaimed_withdraw(amount)?.update_or_kill()? {
+			let _ = frame_system::Pallet::<T>::dec_providers(who).defensive();
+		}
+		// book keep delegation
+		delegation.amount = delegation
+			.amount
+			.checked_sub(&amount)
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+
+		// remove delegator if nothing delegated anymore
+		delegation.update_or_kill(delegator);
+
+		let released = T::Currency::release(
+			&HoldReason::StakingDelegation.into(),
+			delegator,
+			amount,
+			Precision::BestEffort,
+		)?;
+
+		defensive_assert!(released == amount, "hold should have been released fully");
+
+		Self::deposit_event(Event::<T>::Released {
+			agent: who.clone(),
+			delegator: delegator.clone(),
+			amount,
+		});
+
+		Ok(())
+	}
+
+	/// Migrates delegation of `amount` from `source` account to `destination` account.
+	fn do_migrate_delegation(
+		source_delegator: &T::AccountId,
+		destination_delegator: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		let mut source_delegation =
+			Delegators::<T>::get(source_delegator).defensive_ok_or(Error::<T>::BadState)?;
```

Comment:
nit but not sure if `BadState` is the best error here.

Reply:
The pattern I have tried to follow is:
- Public fn does all the sanity checks.
- Internal (pvt) fn assumes the passed parameters are sane.

The delegator checks are already done by the caller and this function assumes it must be a delegator or returns `BadState` defensively. Wdyt?

Reply:
Least assumptions about the call site generally better.